boards: add Wemos D1 R32 (aka ESPDuino-32) board definition

gschorcht opened the pull request:
### Contribution description

This PR adds the board definition for the Wemos D1 R32 also known as ESPDuino-32, the only  only Arduino-sized ESP32 board with Arduino Uno compatible headers.

<img src="https://makerselectronics.com/wp-content/uploads/2022/09/wemos-d1-r32-w-esp32-uno-r3-pinout-1-800x800.jpg.webp" alt="WeMos D1 R32" width="400"/>

### Testing procedure

Compilation in CI should succeed.

### Issues/PRs references

crasbe commented:
I just bought the hardware and once it arrives, I'll give it a test.

In the meantime, you could change the headerguards to `#pragma once` as discussed in https://github.com/RIOT-OS/RIOT/issues/21335 and the `doc.txt` files to `doc.md` as discussed in #21220.

mguetschow commented on boards/esp32-wroom-32/doc.txt:
```diff
@@ -96,8 +96,8 @@ Function        | GPIOs  | Remarks |Configuration
 ADC             | GPIO0, GPIO2, GPIO4, GPIO12, GPIO13,\n GPIO14, GPIO15, GPIO25, GPIO26, GPIO27,\n GPIO32, GPIO33, GPIO34, GPIO35, GPIO36,\n GPIO39 | | see \ref esp32_adc_channels "ADC Channels"
-DAC             | GPIO25, GPIO26 | | \ref esp32_dac_channels "refer"
-PWM_DEV(0)      | GPIO0, GPIO2, GPIO4, GPIO16, GPIO17 | - | \ref esp32_pwm_channels "DAC Channels"
+DAC             | GPIO25, GPIO26 | | \ref esp32_dac_channels \ref esp32_pwm_channels "DAC Channels"
```
```suggestion
DAC             | GPIO25, GPIO26 | | \ref esp32_dac_channels "DAC Channels"
```

mguetschow commented on boards/esp32-wemos-d1-r32/Kconfig:
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2020 HAW Hamburg
```
maybe consider changing the copyright here

gschorcht replied:
Thanks, at typical copy&paste problem. The Kconfig still had to be completely updated :sunglasses:

crasbe commented on boards/esp32-wemos-d1-r32/doc.md:
```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2025 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
```
```suggestion
```

No copyright notices and block comments are required for Markdown files.
(Remember to remove the closing piece of the block comment below as well).

crasbe commented on boards/esp32-wemos-d1-r32/doc.md:
```diff
@@ -0,0 +1,137 @@
+ */
+
+/**
+ * @defgroup    boards_esp32_wemos_d1_r32 Wemos D1 R32 (ESPDuino-32)
+ * @ingroup     boards_esp32
+ * @brief       Support for the Wemos D1 R32 (ESPDuino-32) board
+ * @author      Gunar Schorcht <gunar@schorcht.net>
```
```suggestion
@defgroup    boards_esp32_wemos_d1_r32 Wemos D1 R32 (ESPDuino-32)
@ingroup     boards_esp32
@brief       Support for the Wemos D1 R32 (ESPDuino-32) board
@author      Gunar Schorcht <gunar@schorcht.net>
```

gschorcht replied:
Ah ok, now where it is a Markdown file, C comments are wrong now :sunglasses:

crasbe commented on boards/esp32-wemos-d1-r32/doc.md:
```diff
@@ -0,0 +1,137 @@
+
+## Table of Contents {#esp32_wemos_d1_r32_toc}
+
+1. [Overview](#esp32_wemos_d1_r32_overview)
+2. [Hardware](#esp32_wemos_d1_r32_hardware)
+    1. [MCU](#esp32_wemos_d1_r32_mcu)
+    2. [Board Configuration](#esp32_wemos_d1_r32_board_configuration)
+    3. [Board Pinout](#esp32_wemos_d1_r32_pinout)
+3. [Flashing the Device](#esp32_wemos_d1_r32_flashing)
```
```suggestion
-# [Overview](#esp32_wemos_d1_r32_overview)
-# [Hardware](#esp32_wemos_d1_r32_hardware)
    -# [MCU](#esp32_wemos_d1_r32_mcu)
    -# [Board Configuration](#esp32_wemos_d1_r32_board_configuration)
    -# [Board Pinout](#esp32_wemos_d1_r32_pinout)
-# [Flashing the Device](#esp32_wemos_d1_r32_flashing)
```

You can let Doxygen do the enumeration for you instead of doing it manually.

crasbe commented on boards/esp32-wemos-d1-r32/doc.md:
```diff
@@ -0,0 +1,127 @@
+
+### Board Configuration {#esp32_wemos_d1_r32_board_configuration}
+
+The boad configuration guaranties that the board is compatible with Arduino
```
```suggestion
The board configuration guarantees that the board is compatible with Arduino
```

gschorcht replied:
Good catch, I'm wondering why it wasn't catched by the spell checker in CI?

crasbe replied:
`guaranties` is the valid plural form of `guaranty` and `boad` is simply not in the library: https://github.com/codespell-project/codespell/blob/d59cf4fe68f1d606c23f21a19ecd3828dbe8725e/codespell_lib/data/dictionary.txt#L9789

crasbe commented on boards/esp32-wemos-d1-r32/include/board.h:
```diff
@@ -0,0 +1,80 @@
+ * mode for the GPIO pin has to be GPIO_IN.
+ */
+#ifndef BTN0_INT_FLANK
+#define BTN0_INT_FLANK  GPIO_FALLING
```
```suggestion
#  define BTN0_INT_FLANK  GPIO_FALLING
```

crasbe replied:
Sorry, I forgot to elaborate. Our *official stylekeeper* @maribu establishes the indentations after the # for `#defines` in `#if` statements to make them more readable, especially for multiple `#if`-layers.

Therefore the suggestion.

gschorcht replied:
@maribu I really like indentation for nested `#ifdef ... #endif` directives or a sequence of defines within a conditional compilation. But especially for a bunch simple `#ifndef`s directives used as macro guards with just a single define depending on whether it is already defined or not, indentation leads to more "choppy" code that makes reading harder rather than better. 
```
#ifndef ABDCX
#define ABDCX  x
#endif

#ifndef ABDCY
#define ABDCX  y
#endif

#ifndef ABDCZ
#define ABDCZ  z
#endif
```
vs.
```
#ifndef ABDCX
#  define ABDCX  x
#endif

#ifndef ABDCX
#  define ABDCY  y
#endif

#ifndef ABDCZ
#  define ABDCZ  z
#endif
```
IMHO, copy & paste errors are also easier to catch with the first variant than with the second variant because of the alignment of the identifiers. Personally, I would prefer if the indentation for such conditionals were not mandatory for such a single definition.

crasbe replied:
I was just joking with the "official stylekeeper", but the indentation is indeed part of the Coding Convention https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#indentation-of-preprocessor-directives

But indeed maribu is a lot more experienced with code styling than I am, so maybe he can voice his opinion.

gschorcht replied:
> I was just joking with the "official stylekeeper", but the indentation is indeed part of the Coding Convention https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#indentation-of-preprocessor-directives

I know, I have read it :wink: but sometimes it is worth to discuss some style rules :sunglasses:

maribu replied:
I would favor to just unconditionally stick with two spaces of indent. Style is pretty subjective, e.g. to me the second example with the indent is actually easier to read than the one without.

I would argue that adding exceptions to the coding conventions for short code makes the coding convention more complex and makes it more difficult to the use of tools such as `clang-format` and `uncrustify`, as special exceptions are not easily configured with such tools. Exceptions also tends to cause a lot of discussion (Would two `#define`s within the `#ifndef` ... `#endif` still be considered as short? How about three?).

In `core` @kaspar030 has enforced an "run `uncrustify`, don't discuss style" policy, even for the few cases the output of `uncrustify` not super ideal. I think we ended up spending more time to discuss the content of the code and almost no time on discussing style, which I believe is a good thing.

But anyway, so far we have not strictly enforced the coding convention outside of `core` yet, so I would not start insisting here.

gschorcht replied:
> Style is pretty subjective, e.g. to me the second example with the indent is actually easier to read than the one without.

Totally agree. Maybe I should look for an editor with better syntax highlighting, where macro names in defines and guards are somehow colored :wink: `gnome-textedit` only uses one color for complete preprocessor directives, so that in header files that only contain preprocessor directives, there is in fact no highlighting :worried:

crasbe commented on boards/esp32-wemos-d1-r32/include/periph_conf.h:
```diff
@@ -0,0 +1,117 @@
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ */
+
+#ifndef PERIPH_CONF_H
+#define PERIPH_CONF_H
+
```
```suggestion
```

Looks like you found a cornercase in the headerguards check 😅
I'll add a PR to cover that.

crasbe replied:
This commit is for another ESP32 board https://github.com/RIOT-OS/RIOT/pull/21479/commits/4e3bb011cbf3c1a7599e83a2a657cd506c586c18 ?

The `boards/esp32-wemos-d1-r32/include/periph_conf.h` is still unchanged.

Did you `cd` into the wrong directory? 😅

gschorcht replied:
I just mixed some work when I was switching between differen branches :see_no_evil:

crasbe commented on boards/esp32-wemos-d1-r32/include/gpio_params.h:
```diff
@@ -0,0 +1,56 @@
+/**
+ * @brief   Button configuration
+ */
+static const saul_gpio_params_t saul_gpio_params[] =
+{
+    {
+        .name = "BOOT",
+        .pin = BTN0_PIN,
+        .mode = BTN0_MODE,
+        .flags = SAUL_GPIO_INVERTED
+    },
+};
+
+/**
+ * @brief   LED configuration
+ */
+static const saul_gpio_params_t saul_gpio_params[] =
+{
+    {
+        .name = "LED",
+        .pin = LED0_PIN,
+        .mode = GPIO_OUT,
+        .flags = (SAUL_GPIO_INVERTED | SAUL_GPIO_INIT_CLEAR)
+    }
+};
```
This should probably be one struct?

The compiler throws the following errors when building `examples/basic/default`:
```
In file included from /home/cbuec/RIOTstuff/riot-gettingstarted/RIOT/drivers/saul/init_devs/auto_init_saul_gpio.c:25:
/home/cbuec/RIOTstuff/riot-gettingstarted/RIOT/boards/esp32-wemos-d1-r32/include/gpio_params.h:33:16: error: 'BTN0_PIN' undeclared here (not in a function)
   33 |         .pin = BTN0_PIN,
      |                ^~~~~~~~
/home/cbuec/RIOTstuff/riot-gettingstarted/RIOT/boards/esp32-wemos-d1-r32/include/gpio_params.h:34:17: error: 'BTN0_MODE' undeclared here (not in a function)
   34 |         .mode = BTN0_MODE,
      |                 ^~~~~~~~~
/home/cbuec/RIOTstuff/riot-gettingstarted/RIOT/boards/esp32-wemos-d1-r32/include/gpio_params.h:42:33: error: redefinition of 'saul_gpio_params'
   42 | static const saul_gpio_params_t saul_gpio_params[] =
      |                                 ^~~~~~~~~~~~~~~~
/home/cbuec/RIOTstuff/riot-gettingstarted/RIOT/boards/esp32-wemos-d1-r32/include/gpio_params.h:29:33: note: previous definition of saul_gpio_params' with type 'const saul_gpio_params_t[1]'
   29 | static const saul_gpio_params_t saul_gpio_params[] =
      |                                 ^~~~~~~~~~~~~~~~
make[3]: *** [/home/cbuec/RIOTstuff/riot-gettingstarted/RIOT/Makefile.base:157: /home/cbuec/RIOTstuff/riot-gettingstarted/RIOT/examples/basic/default/bin/esp32-wemos-d1-r32/saul_init_devs/auto_init_saul_gpio.o] Error 1
make[2]: *** [/home/cbuec/RIOTstuff/riot-gettingstarted/RIOT/Makefile.base:31: ALL--/home/cbuec/RIOTstuff/riot-gettingstarted/RIOT/drivers/saul/init_devs] Error 2
```

gschorcht replied:
> This should probably be one struct?

Yeah, of course. But how could the compilation in Murdock succeed with this error, I'm a bit confused.

Anyway, since the board has no button, I will just drop the first structure. The problem was probably caused by copying and pasting these structures from different boards, sorry.

crasbe replied:
The CI only compiles for certain boards and this board is not one of them. Therefore the error would've shown only in the full build when the tests and examples are compiled for all boards, including this one.

crasbe commented on boards/esp32-wemos-d1-r32/include/gpio_params.h:
```diff
@@ -0,0 +1,43 @@
+        .name = "LED",
+        .pin = LED0_PIN,
+        .mode = GPIO_OUT,
+        .flags = (SAUL_GPIO_INVERTED | SAUL_GPIO_INIT_CLEAR)
```
```suggestion
        .flags = (SAUL_GPIO_INIT_CLEAR)
```

With the inverted flag, the LED is off when writing 1 and on when writing 0.
Furthermore, the LED is on on startup.

crasbe reviewed (approved):
You can squash the fixups. Please also address the suggestion from @mguetschow.

I tested the board locally with `examples/basic/default`, `tests/periph/adc` and `tests/periph/pwm` and everything seems to work as it should.

One thing that is still on my mind is the removal of the `arduino_board_common.h` in this PR. Maybe #21484 would be a better home for that, since it's already an "ESP32 Arduino Cleanup" PR?

gschorcht commented:
> You can squash the fixups. Please also address the suggestion from @mguetschow.

These suggestions were already addressed. 

> One thing that is still on my mind is the removal of the `arduino_board_common.h` in this PR. Maybe #21484 would be a better home for that, since it's already an "ESP32 Arduino Cleanup" PR?

I moved the commit to #21484 and squashed.

Thanks for reviewing and testing.

gschorcht commented:
Hm, compilation failed in Murdock because `tests/drivers/sps30/Makefile` defines the macro `I2C0_SPEED`
```
CFLAGS+=-DI2C0_SPEED=I2C_SPEED_NORMAL
```
It is not yet clear for what purpose the macro was defined. It seems that it is not used in the driver code. Maybe it's just a leftover from the time when `i2c_init` took the I2C clock frequency as a parameter and the application could define which I2C clock frequency to use. On the other hand, the I2C API was changed in 2018, while the application was added with PR #13256 in 2020 :thinking:

In the description of PR #13256, it is pointed out that the SPS30 only works at normal speed. Maybe the definition of `I2C0_SPEED` in the Makefile was an attempt to define the speed at least for ESP32x boards, where this macro is actually used, but Fast Speed is the default.

I now conditionally define `I2C0_SPEED` in the board configuration.

gschorcht commented:
I had to modify the comments and mapping for Arduino DAC pins with commit 1e274e194fbc0755dd004126b4e425b10f7568a6.

gschorcht commented:
@crasbe With commit 1e274e194fbc0755dd004126b4e425b10f7568a6 I tried to changed the order of the Arduino DAC pins so that Arduino pins `DAC0` and `DAC1` are mapped consistently to `DAC_LINE(0)` and `DAC_LINE(1)`, respectively.
https://github.com/RIOT-OS/RIOT/blob/694e4a449fdb2c1c958deb1c2012bd558f877d7b/boards/esp32-wemos-d1-r32/include/periph_conf.h#L45
https://github.com/RIOT-OS/RIOT/blob/694e4a449fdb2c1c958deb1c2012bd558f877d7b/boards/esp32-wemos-d1-r32/include/arduino_iomap.h#L102-L103
However, this results in the Arduino pin D2 being the `DAC1` and D3 the `DAC0`. That is, while the digital Arduino pins D2 and D3 are labelled in ascending order, the DAC pins are labelled in descending order. With other words you have D2=`DAC1` and D3=`DAC0` which might be a little bit confusing.
https://github.com/RIOT-OS/RIOT/blob/694e4a449fdb2c1c958deb1c2012bd558f877d7b/boards/esp32-wemos-d1-r32/include/arduino_iomap.h#L40-L41

Now I'm wondering whether it wouldn't be better to change the order in `periph_conf.h` to
```
#define DAC_GPIOS   { GPIO26, GPIO25 } 
```
so that we could map D2=`DAC0` to DAC_LINE(0) and D3=`DAC1` to `DAC_LINE(1)`.

On the other hand, DAC pins are something that is not known in the Arduino world and it probably doesn't matter how you map them to the DAC lines. What do you think?

crasbe commented:
> What do you think?

You are probably overthinking it :D
I don't have an opinion on it and I agree that it probably doesn't matter.

gschorcht commented:
Ok, since we live in the RIOT bubble where Arduino sketches are just a nice but very special gadget, I would say that it is more important to map the ESP32 GPIOs to the `DAC_LINE` in ascending order, also because the headers of the board are not labelled `D0` ... `D13`, but with `GPIOx`.

So I would have to squash the fixups.

gschorcht commented:
May I enable `Auto-merge`?

crasbe commented:
Any time :)

gschorcht commented:
@crasbe Many thanks for reviewing :wink:
